### ISSUE TITLE
[UBSAN] TkDetLayers/src/TIBRing.cc warning: 'D.56128.closestIndex' may be used uninitialized 

### ISSUE THREAD

**Issue by iarspider** [cms-sw/cmssw#40925]
```
>> Building shared library tmp/el8_amd64_gcc11/src/RecoTracker/TkDetLayers/src/RecoTrackerTkDetLayers/libRecoTrackerTkDetLayers.so
(...)
In member function 'computeCrossings',
    inlined from 'groupedCompatibleDetsV' at .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:115:31:
  .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: warning: 'D.56128.closestIndex' may be used uninitialized [-Wmaybe-uninitialized]
   210 |     return SubRingCrossings();
      |                             ^
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc: In member function 'groupedCompatibleDetsV':
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: note: '<anonymous>' declared here
  210 |     return SubRingCrossings();
      |                             ^
In member function 'computeCrossings',
    inlined from 'groupedCompatibleDetsV' at .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:115:31:
  .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: warning: 'D.56128.nextIndex' may be used uninitialized [-Wmaybe-uninitialized]
   210 |     return SubRingCrossings();
      |                             ^
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc: In member function 'groupedCompatibleDetsV':
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: note: '<anonymous>' declared here
  210 |     return SubRingCrossings();
      |                             ^
In member function 'computeCrossings',
    inlined from 'groupedCompatibleDetsV' at .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:115:31:
  .../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: warning: 'D.56128.nextDistance' may be used uninitialized [-Wmaybe-uninitialized]
   210 |     return SubRingCrossings();
      |                             ^
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc: In member function 'groupedCompatibleDetsV':
.../src/RecoTracker/TkDetLayers/src/TIBRing.cc:210:29: note: '<anonymous>' declared here
  210 |     return SubRingCrossings();
      |                             ^
```

Full log: [link](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc11/CMSSW_13_1_UBSAN_X_2023-03-01-2300/RecoTracker/TkDetLayers)

**Comment by iarspider** [comment]
assign reconstruction

**Comment by slava77** [comment]
@iarspider 
The log doesn't show where the use case of uninitialized value can happen.
The coding design is to use `isValid_` flag to check if the rest is meaningful.
Is this warning just a case of "you have to initialize everything"? This comes at a cost and should not be required.

Also, this coding pattern repeats in `TIDRing::computeCrossings`, but doesn't appear in the build logs. Does the build stop on the first compile unit with warnings?

**Comment by slava77** [comment]
type tracking

**Comment by makortel** [comment]
I had played last week a bit, and I was able to get rid of the warning by replacing the `isValid_` flag with the use of `std::optional`. https://github.com/cms-sw/cmssw/pull/41164 shows how it looks. I did not do any performance evaluation.

**Comment by iarspider** [comment]
Fixed with #41164 .
